### PR TITLE
Add contrib importance pie chart

### PR DESCRIPTION
**Pull request by christinaexyou**
This PR is part 2 of addressing issue #114 by making the following changes:
1. Creates a pie chart depicting contributor importance 
2. Updates `queries/contributors_query.py` to query for contributor alias emails and returns the dataframe with an `email` column included
3. Updates chaoss/chaoss.py

### CODE REVIEW

File: pages/chaoss/visualizations/contrib_importance.py
Comment (cdolfi) [inline pages/chaoss/visualizations/contrib_importance.py:207]: What about doing the bot filtering by the login? Not sure how many bots actually have emails linked to the account beyond a "githubnoreply" one. Also would be able to get it from the contributors actions query without any changes. I am also seeing if augur get bot tags (https://github.com/3scale/3scale-operator/pull/860) example of a bot with a bot tags, will luk. @JamesKunstle thoughts?

File: pages/chaoss/visualizations/contrib_importance.py
Comment (cdolfi) [inline pages/chaoss/visualizations/contrib_importance.py:20]: change to "contrib-importance-pie" or something like it to make the naming descriptively distinct to the line graph (same for the file name)

File: pages/chaoss/visualizations/contrib_importance.py
Comment (cdolfi) [inline pages/chaoss/visualizations/contrib_importance.py:27]: A little more descriptive title maybe, @JamesKunstle thoughts?

**Review by cdolfi — CHANGES_REQUESTED** [review]
A few requested changes. Going to look into the bot tagging!

File: pages/chaoss/visualizations/contrib_importance_pie.py
Comment (cdolfi) [inline pages/chaoss/visualizations/contrib_importance_pie.py:37]: change to be by log in not email

File: queries/contributors_query.py
Comment (cdolfi) [inline queries/contributors_query.py:12]: Drop this file from being changed. Should be the same as before

File: pages/chaoss/visualizations/contrib_importance_pie.py
Comment (cdolfi) [inline pages/chaoss/visualizations/contrib_importance_pie.py:203]: Should this be in the same format as line 200? I thought isin would require that df.login (in this example) string is in the patterns list, no the subset

**Comment by christinaexyou** [comment]
Changes made:

1. Addressed no data issue for PR Review and Issue Open 
2. Search for potential bots in login rather than email - this is mostly just a temp fix
3. Revert contributors_query.py to the most recent update in the dev branch

File: pages/chaoss/visualizations/contrib_importance_pie.py
Comment (cdolfi) [reply pages/chaoss/visualizations/contrib_importance_pie.py:37]: Add details on the cntrb ids, (anonymized logins)

File: pages/chaoss/visualizations/contrib_importance_pie.py
Comment (cdolfi) [inline pages/chaoss/visualizations/contrib_importance_pie.py:236]: Add in comment that names could be changed to "login" to un- anonymize

File: pages/chaoss/visualizations/contrib_importance_pie.py
Comment (cdolfi) [inline pages/chaoss/visualizations/contrib_importance_pie.py:239]: Take out hole to be able to see splits better and match format of other pie charts

File: pages/chaoss/visualizations/contrib_importance_pie.py
Comment (cdolfi) [inline pages/chaoss/visualizations/contrib_importance_pie.py:27]: @JamesKunstle Curious to see if you have any ideas on naming of this graph, the line, and then the action types viz I made last week. Trying to think if there is any better naming to make them distinct or if its not really possible/ big deal

File: pages/chaoss/visualizations/contrib_importance_pie.py
Comment (cdolfi) [inline pages/chaoss/visualizations/contrib_importance_pie.py:246]: Add "Contrib Id" before label or another identifier

**Review by cdolfi — CHANGES_REQUESTED** [review]
Some small changes requested. Its getting close!

**Comment by cdolfi** [comment]
@christinaexyou is this meant to be closed?

File: pages/chaoss/visualizations/contrib_importance_pie.py
Comment (cdolfi) [inline pages/chaoss/visualizations/contrib_importance_pie.py:40]: Add "Note: Some commits have a contributor ID as none if there is no github account associated with the email used to commit"

**Review by cdolfi — CHANGES_REQUESTED** [review]
One small additional review comment. I resolved all of the prior comments that have been addressed. Once the rest are handled this will be good to go!

File: pages/chaoss/visualizations/contrib_importance_pie.py
Comment (christinaexyou) [reply pages/chaoss/visualizations/contrib_importance_pie.py]: Added the following comment in line 242: `# can be replaced with login to unanonymize`

File: pages/chaoss/visualizations/contrib_importance_pie.py
Comment (christinaexyou) [reply pages/chaoss/visualizations/contrib_importance_pie.py:209]: Double checked and isin checks if df.login is an exact match and NOT substring matches to any element in the patterns list. Edited my comment  in line 208: `# remove rows where the login column value is in patterns - requires the login string to be in patterns`

File: pages/chaoss/visualizations/contrib_importance_pie.py
Comment (christinaexyou) [reply pages/chaoss/visualizations/contrib_importance_pie.py]: Added note - replaced "commit" to "contributions" to generalize

**Comment by christinaexyou** [comment]
The following changes were made:
1. Corrected original comment about `isin` for removing bots - it looks for exact matches between the login and the elements in patterns NOT substring matches
2. Updated graph info with a note about contributors that are ID'ed as None

File: pages/chaoss/visualizations/contrib_importance_pie.py
Comment (cdolfi) [reply pages/chaoss/visualizations/contrib_importance_pie.py:209]: So this should be changed to .str.contains(patterns) as we are looking for substrings and that is what the search bar is prompting for

**Review by cdolfi — CHANGES_REQUESTED** [review]
Updated review comments

File: pages/chaoss/visualizations/contrib_importance_pie.py
Comment (cdolfi) [inline pages/chaoss/visualizations/contrib_importance_pie.py:212]: Use ~ not ==False and use the following logic: https://www.skytowner.com/explore/getting_rows_where_column_value_contains_any_substring_in_a_list_in_pandas_dataframe

**Review by cdolfi — CHANGES_REQUESTED** [review]
One last thing and then just waiting on @JamesKunstle 's review

File: pages/chaoss/visualizations/contrib_importance_pie.py
Comment (christinaexyou) [reply pages/chaoss/visualizations/contrib_importance_pie.py]: Included `na=False` clause in mask variable since there are some nan/empty login column values.

Comment (cdolfi) [reply pages/chaoss/visualizations/contrib_importance_pie.py]: Have you specifically check if using ~ works for nan and empty values? Every check ive done shows that it would

Comment (christinaexyou) [reply pages/chaoss/visualizations/contrib_importance_pie.py]: When I just do ~, it returns `TypeError: bad operand type for unary -: 'NoneType' ` so I referenced this post: https://stackoverflow.com/questions/52297740/typeerror-bad-operand-type-for-unary-float

**Review by cdolfi — COMMENTED** [review]
lgtm! Ready to merge once @JamesKunstle has a chance to take a look at the few comments left

File: pages/chaoss/visualizations/contrib_importance_pie.py
Comment (JamesKunstle) [inline pages/chaoss/visualizations/contrib_importance_pie.py:44]: Here's a slight reframe suggestion:

"For a given kind of contribution (e.g. Commit, PR, Issue), visualizes the proportional share of the top 'k' anonymous contributors, aggregating the remaining contributors as "other".

e.g. Contributor 'A' opens the most PRs of all contributors, accounting for 1/5 of all PRs. If 'k' = 1, then the chart will have one slice for Contributor 'A' accounting for 1/5 of the area, with the remaining 4/5 representing all other contributors.

Optionally, filter out contributors who have 'bot' in their login username.

Note: Some commits may have a Contributors ID as 'none' if no Github account is associated with the email that the contributor committed as."

File: pages/chaoss/visualizations/contrib_importance_pie.py
Comment (JamesKunstle) [inline pages/chaoss/visualizations/contrib_importance_pie.py:226]: in Pandas, it's preferred to use Series methods rather than DataFrame indexing, so I believe this line would be replaced with:

`t_sum = df[action_type].sum()`

[I think.](https://www.geeksforgeeks.org/python-pandas-series-sum/#)

File: pages/chaoss/visualizations/contrib_importance_pie.py
Comment (JamesKunstle) [inline pages/chaoss/visualizations/contrib_importance_pie.py:229]: Same kind of semantic Pandas thing here:

`df = df.head(top_k).drop('cntrbs')`

https://sparkbyexamples.com/pandas/get-first-n-rows-of-pandas/#:~:text=DataFrame.,rows%20from%20the%20given%20DataFrame.

File: pages/chaoss/visualizations/contrib_importance_pie.py
Comment (JamesKunstle) [inline pages/chaoss/visualizations/contrib_importance_pie.py:235]: Same kind of semantic Pandas thing, again, because I'm picky:

`df_sum = df[action_type].sum()`

I assume this is for the action column.

File: pages/chaoss/visualizations/contrib_importance_pie.py
Comment (JamesKunstle) [inline pages/chaoss/visualizations/contrib_importance_pie.py:238]: so me, adding a row should be refactored to:

`df = df.append({"cntrb_id": "Other", action_type: str(t_sum - df_sum)})`

because the `df.loc[len(df.index)]` hack is difficult for a novice to read.

**Review by JamesKunstle — CHANGES_REQUESTED** [review]
Minor semantic Pandas updates requested but nothing technical- this seems very good.

I added a recommendation for the 'about' section and the title- please let me know what you think.

File: pages/chaoss/visualizations/contrib_importance_pie.py
Comment (christinaexyou) [reply pages/chaoss/visualizations/contrib_importance_pie.py:44]: What about "Top {k} Contributors by {action}" where k and action are dynamically changing based on the user input ? For the pie chart, I'm not explicitly calculating bus factor - just visualizing the data as is. I like the reframe suggestion - will update `Graph Info`.

Comment (cdolfi) [reply pages/chaoss/visualizations/contrib_importance_pie.py:44]: @JamesKunstle

Comment (JamesKunstle) [reply pages/chaoss/visualizations/contrib_importance_pie.py:44]: @christinaexyou @cdolfi Yeah "Top {k} Contributors by Number of {action}" might be slightly clearer but we're within a razors margin of a perfect title, to me, either way, so I'm down for your preference.

**Comment by christinaexyou** [comment]
@cdolfi @JamesKunstle 
Changes made:

1. Added functionality user to create new keywords to filter out contributors by 
2. Create dynamically changing title
3. Updated `Graph Info` description
4. Edited Pandas logic

File: pages/chaoss/visualizations/contrib_importance_pie.py
Comment (cdolfi) [reply pages/chaoss/visualizations/contrib_importance_pie.py:45]: lgtm on info and graph title, will leave for @JamesKunstle

**Review by cdolfi — APPROVED** [review]
@JamesKunstle to do final review but lgtm

**Review by JamesKunstle — APPROVED** [review]
I think this is done. Good work!

**Pull request merged**
